lsps0/server.go

```diff
@@ -0,0 +1,273 @@
+type serviceInfo struct {
+	// Contains the implementation for the methods in this service.
+	serviceImpl interface{}
+	methods     map[string]*MethodDesc
+}
+
+type methodInfo struct {
+	service *serviceInfo
+	method  *MethodDesc
+}
+
+type Server struct {
+	mu       sync.Mutex
+	serve    bool
+	services map[string]*serviceInfo
+	methods  map[string]*methodInfo
+}
+
+func NewServer() *Server {
+	return &Server{
+		services: make(map[string]*serviceInfo),
+		methods:  make(map[string]*methodInfo),
+	}
+}
+
+func (s *Server) Serve(lis lightning.CustomMsgClient) error {
+	s.mu.Lock()
+	if s.serve {
+		return ErrAlreadyServing
+	}
+	s.serve = true
+	s.mu.Unlock()
+
+	defer func() {
+		s.mu.Lock()
+		s.serve = false
+		s.mu.Unlock()
+	}()
+
+	for {
+		msg, err := lis.Recv()
+		if err != nil {
+			if err == context.Canceled {
+				log.Printf("lsps0: lis got canceled, stopping.")
+				return err
+			}
+
+			log.Printf("lsps0 Serve(): Recv() err != nil: %v", err)
+			<-time.After(time.Second)
+			continue
+		}
+
+		// Ignore any message that is not an lsps0 message
+		if msg.Type != Lsps0MessageType {
+			continue
+		}
+
+		// Make sure there are no 0 bytes
+		if slices.Contains(msg.Data, 0x00) {
+			log.Printf("UNUSUAL: Got custom message containing 0 bytes from peer '%s'.", msg.PeerId)
+			go sendError(lis, msg, nil, status.New(codes.ParseError, BadMessageFormatError))
+			continue
+		}
+
+		req := new(jsonrpc.Request)
+		err = json.Unmarshal(msg.Data, req)
+		if err != nil {
+			log.Printf("UNUSUAL: Failed to unmarshal custom message from peer '%s': %v", msg.PeerId, err)
+			go sendError(lis, msg, nil, status.New(codes.ParseError, BadMessageFormatError))
+			continue
+		}
+
+		if req == nil {
+			log.Printf("UNUSUAL: req == nil after unmarshal custom message from peer '%s': %v", msg.PeerId, err)
+			go sendError(lis, msg, nil, status.New(codes.ParseError, BadMessageFormatError))
+			continue
+		}
+
+		if req.JsonRpc != jsonrpc.Version {
+			log.Printf("UNUSUAL: jsonrpc version is '%s' in custom message from peer '%s': %v", req.JsonRpc, msg.PeerId, err)
+			go sendError(lis, msg, req, status.Newf(codes.InvalidRequest, "Expected jsonrpc %s, found %s", jsonrpc.Version, req.JsonRpc))
+			continue
+		}
+
+		m, ok := s.methods[req.Method]
+		if !ok {
+			log.Printf("UNUSUAL: peer '%s' requested method '%s', but it does not exist.", msg.PeerId, req.Method)
+			go sendError(lis, msg, req, status.New(codes.MethodNotFound, "method not found"))
+			continue
+		}
+
+		df := func(v interface{}) error {
+			if err := json.Unmarshal(req.Params, v); err != nil {
+				return status.Newf(codes.InvalidParams, "invalid params").Err()
+			}
+
+			return nil
+		}
+
+		// NOTE: The handler is being called synchonously. There's an option to
+		// do this in a goroutine instead. Also, there's the option to put the
+		// goroutine in the method desc for specific methods instead.
+		r, err := m.method.Handler(m.service.serviceImpl, context.TODO(), df)
```

Comment:
Yes, I think it should be called in go routine otherwise the requests will block each other (in contrary to the grpc server behavior).

Reply:
It now runs in a goroutine. On top of this, there's a mechanism to make sure there are not too many simultaneous requests.